src/python/pants/engine/goal.py

```diff
@@ -0,0 +1,40 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from pants.option.optionable import Optionable
+from pants.option.scope import ScopeInfo
+from pants.util.meta import AbstractClass, classproperty
+
+
+class Goal(Optionable, AbstractClass):
+  """A CLI goal whch is implemented by a `@console_rule`.
+  
+  This abstract class should be subclassed and given a `Goal.name` that it will be referred to by
+  when invoked from the command line. The `Goal.name` also acts as the options_scope for the `Goal`.
+  """
+
+  # Subclasser-defined. See the class pydoc.
+  name = None
+
+  options_scope_category = ScopeInfo.GOAL
+
+  @classproperty
+  def options_scope(cls):
+    if not cls.name:
+      # TODO: Would it be unnecessarily magical to have `cls.__name__.lower()` always be the name?
```

Comment:
We'd want snake_case, rather than lower, and probably want it to be overridable? Or maybe not overridable... It _is_ a bit magical, but it's magic I kind of like...

Reply:
I think that would be a fine default. It should probably be overridable though.
Devil's advocate: Should Goals necessarily be optionable at all? Do we want that coupling? Why not let authors define subsystems where needed?